### PR TITLE
[utils] job manager/queue fixes

### PR DESCRIPTION
**Pull request by ksooo**
Two fixes here:

1) Memory leak. Jobs NOT successfully added to CJobManager, because the manager is not running, were never deleted. I checked every calling code and found, that no callee inspects the return value of AddJob and all jobs are actually leaked. The callee not deleting the job is fine as the contract is that the callee hands over ownership of the job it has created to CJobManager. Now, AddJob deletes the job passed to it in case it cannot be added. I updated the documentation accordingly. BTW, CJobQueue::AddJob did not have the memory leak, because it was already deleting the passed job in case of error. I updated CJobQueue's documentation to be clear on that behavior.

2) Fixed access to freed memory thus crash happened after CJobManager::CancelJobs was called (on Kodi exit). The two annotated screen shots try to explain what happened exactly:

Screenshot 2021-05-04 at 15 47 32
Screenshot 2021-05-04 at 15 48 46

As I can easily reproduce both the crash and the memory leak, I did extensive runtime tests and know that the two commits of this PR do fix both abovementioned problems.

@phunkyfish @lrusak due to lack of a maintainer for the CJob* classes, could you please take a look?

### CODE REVIEW

**Comment by ksooo** [comment]
```diff
diff --git a/xbmc/utils/Job.h b/xbmc/utils/Job.h
index 00a1e67abc..9c8664aacf 100644
--- a/xbmc/utils/Job.h
+++ b/xbmc/utils/Job.h
@@ -77,7 +77,7 @@ public:
    \param job the job that has been aborted.  The job will be destroyed after this function returns
    \sa CJobManager and CJob
    */
-  virtual void OnJobAbort(unsigned int jobID, CJob* job) {};
+  virtual void OnJobAbort(unsigned int jobID, CJob* job){};
 };
 
 class CJobManager;
```

Really? No.

File: xbmc/utils/Job.h [hunk 1]
```diff
@@ -64,6 +65,19 @@ class IJobCallback
    \sa CJobManager and CJob
    */
   virtual void OnJobProgress(unsigned int jobID, unsigned int progress, unsigned int total, const CJob *job) {};
+
+  /*!
+   \brief An optional callback function used when a job will be aborted before it was executed.
+
+   OnJobAbort is called whenever a job gets aborted before it was executed. Job's DoWork method
+   will not be called, OnJobComplete will not be called.  The job instance will be destroyed right
```
Comment (lrusak) [inline xbmc/utils/Job.h:73]: It looks to me that `OnJobAbort` directly calls `OnJobComplete`. Isn't that a contradiction to this documentation?

File: xbmc/utils/JobManager.cpp [hunk 2]
```diff
@@ -122,19 +127,21 @@ bool CJobQueue::AddJob(CJob *job)
...
   CSingleLock lock(m_section);
-  if (m_jobQueue.size() && m_processing.size() < m_jobsAtOnce)
+  while (m_jobQueue.size() && m_processing.size() < m_jobsAtOnce)
   {
-    CJobPointer &job = m_jobQueue.back();
+    CJobPointer& job = m_jobQueue.back();
     job.m_id = CJobManager::GetInstance().AddJob(job.m_job, this, m_priority);
-    m_processing.push_back(job);
+    if (job.m_id > 0)
+      m_processing.push_back(job);
```
Comment (lrusak) [inline xbmc/utils/JobManager.cpp:143]:
```suggestion
      m_processing.emplace_back(job);
```

File: xbmc/utils/Job.h [hunk 3]
```diff
@@ -64,6 +65,19 @@ class IJobCallback
    \sa CJobManager and CJob
    */
   virtual void OnJobProgress(unsigned int jobID, unsigned int progress, unsigned int total, const CJob *job) {};
+
+  /*!
+   \brief An optional callback function used when a job will be aborted before it was executed.
+
+   OnJobAbort is called whenever a job gets aborted before it was executed. Job's DoWork method
```
Comment (lrusak) [inline xbmc/utils/Job.h:72]: Would this ever be used in the case that a running job needs to be aborted?

**Review by lrusak — COMMENTED** [review]
I don't have any knowledge of this code so I just left a couple comments about what I directly saw in the changes made.

File: xbmc/utils/Job.h (hunk 1 above)
Comment (ksooo) [reply xbmc/utils/Job.h:73]: No no, this is just a detail for CJobQueue implementing both call backs the same way, not  part of the general concept. CJobQueue has to do the same, regardless the job was executed or aborted before execution - namely to remove the job from its own processing queue and to continue processing by adding the next job(s) to the job manager. Other use cases might have very differently looking implementations for abort vs. completion of execution.

File: xbmc/utils/JobManager.cpp (hunk 2 above)
Comment (ksooo) [reply xbmc/utils/JobManager.cpp:143]: Ah yes, will change that.

File: xbmc/utils/Job.h (hunk 3 above)
Comment (ksooo) [reply xbmc/utils/Job.h:72]: No, if a running job shall be canceled, this can be done via OnJobProgress.

File: xbmc/utils/Job.h (hunk 1 above)
Comment (lrusak) [reply xbmc/utils/Job.h:78]: Thank you for the clarification :+1:

File: xbmc/utils/Job.h (hunk 3 above)
Comment (lrusak) [reply xbmc/utils/Job.h:77]: Thanks for the explanation.

Comment (ksooo) [reply xbmc/utils/Job.h:77]: > Would this ever be used in the case that a running job needs to be aborted?

 On second thought, you are right. This new callback has also be called when job manager cancels a running job. I will change the PR accordingly.

**Review by phunkyfish — COMMENTED** [review]
Just some minors in the documentation.

I'm building now to do a runtime test.

File: xbmc/utils/Job.h [hunk 4]
```diff
@@ -51,6 +52,19 @@ class IJobCallback
    */
   virtual void OnJobComplete(unsigned int jobID, bool success, CJob *job)=0;
 
+  /*!
+   \brief An optional callback function used when a job will be aborted.
+
+   OnJobAbort is called whenever a job gets aborted before being executed or while being execuded.
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/utils/Job.h:58]: Thanks, I will change that.

File: xbmc/utils/Job.h [hunk 5]
```diff
@@ -106,9 +125,9 @@ class CJob
   /*!
    \brief Destructor for job objects.
 
-   Jobs are destroyed by the CJobManager after the OnJobComplete() callback is complete.
-   CJob subclasses  should therefore supply a virtual destructor to cleanup any memory allocated by
-   complete or cancelled jobs.
+   Jobs are destroyed by the CJobManager after the OnJobComplete() or OnJobAbort() callback is
+   complete.  CJob subclasses  should therefore supply a virtual destructor to cleanup any memory
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/utils/Job.h:129]: Yep, will fix.

**Comment by ksooo** [comment]
As the crash is really hard to reproduce, here is a proof that the fix is actually needed and that it is working as intended:

Screenshot 2021-05-07 at 08 32 32

**Comment by arnova** [comment]
LGTM. Nice catch! Unbelievable this was creeping in XBMC's/Kodi's code for ages.

**Review by phunkyfish — APPROVED** [review]
Looks good now

**Pull request merged**